### DISCUSSION TITLE
protoc version

### DISCUSSION THREAD

**Discussion by jjh42** [rules-proto-grpc/rules_proto_grpc#374] [Q&A] [answered]
### Description

Is there some way to easily select the version of protoc that is used.

The issue is that I need to use protoc that generates python version 4 files
https://protobuf.dev/support/version-support/
since almost all python dependencies are still on that version.

But I can't figure out how to do this.

**Comment by sschnug** [comment, marked as answer]
I also currently look into the new bzlmod-based version of rules-proto-grpc and the protobuf cross-runtime guarantee topic.

Consider trying out the following:

- Follow the setup in https://rules-proto-grpc.com/en/stable/lang/cpp.html#installation -> `protoc.toolchain`
  - C++ has no cross-runtime guarantees and therefore this explicit setting is extremely important (and maybe for this reason it's only to be found in the c++ example!) 
 
Now is this respected in other toolchains? -> Yes!

- `version = "v27.1"` -> produces python with header docstrings containing `# Protobuf Python Version: 5.27.1`
  - proof of correct core <-> language mapping: https://github.com/protocolbuffers/protobuf/blob/v27.1/version.json
- `version = "v26.0"` -> `# Protobuf Python Version: 5.26.0`
  - https://github.com/protocolbuffers/protobuf/blob/v26.0/version.json
- `version = "v25.4"` -> `# Protobuf Python Version: 4.25.4` **!!!!!**
  - https://github.com/protocolbuffers/protobuf/blob/v25.4/version.json

Now the versions you can pick from must be pre-registered afaik (aka it fails for example with `25.5`). It seems the source of truth here is https://github.com/aspect-build/toolchains_protoc/blob/main/protoc/private/versions.bzl
